### ISSUE TITLE
MPI_Comm_set/get_info and relatives - need to check for MPI 4.0 compliance

### ISSUE THREAD

**Issue by hppritcha** [open-mpi/ompi#9207]
In the MPI 4.0 standard, additional verbiage was added to the descriptions of ```MPI_COMM_SET_INFO``` and ```MPI_COMM_GET_INFO```.  It should be verified that Open MPI is adhering to this new language.

In particular, for ```MPI_COMM_SET_INFO``` the following sentences were added to the description (see 7.4 page 347 of the MPI 4.0 standard):

```
This operation has no effect on previously set or defaulted
hints that are not specified by info. It also has no effect on previously set or defaulted
hints that are specified by info, but are ignored by the MPI implementation in this call to
MPI_COMM_SET_INFO.
```

and for ```MPI_COMM_GET_INFO``` the following verbiage was added to the description of this function in the MPI 4.0 standard

```
An MPI implementation is required to return all hints that are
supported by the implementation and have default values specified; any user-supplied hints
that were not ignored by the implementation; and any additional hints that were set by
the implementation.
```

**Comment by devreal** [comment]
The current implementation is not compliant as it does not  hand out modifications made through `MPI_Comm_info_set` (IIRC). The reason is the last parameter in the call to `opal_info_dup_mode`: https://github.com/open-mpi/ompi/blob/master/opal/util/info.c#L289

There is plenty of code that deals with potentially different interpretations of the standard, e.g., the whole saving of existing values when a new one is set using the `OPAL_INFO_SAVE_PREFIX`. IMO, with the clarifications in 4.0 there is no ambiguity and we should get rid of this, as suggested here: https://github.com/open-mpi/ompi/issues/8505

I'll be happy to take that up (since I have fiddled with the info infrastructure recently) if we can agree that saving old info values is not needed (anymore). I poked around to try and understand why the code is what it is but I didn't get more than "there once was a discussion and this was the outcome"...
